### PR TITLE
Enable interleaved sliding_window for gemma3

### PR DESCRIPTION
**Pull request by jiminha**
Enable interleaved sliding_window for gemma3 
 - apply sliding_window mask for prompt attention
 - apply sliding_window for decode block information 
 
 Without this change, prompts longer than size of sliding_window have an accuracy issue. 
 

 vllm (pretrained=google/gemma-3-4b-it,tensor_parallel_size=1,distributed_executor_backend=mp,trust_remote_code=true,max_model_len=4096,use_v2_block_manager=True,dtype=bfloat16,max_num_seqs=128), gen_kwargs: (None), limit: None, num_fewshot: 8, batch_size: 128
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     8|exact_match|↑  |0.7574|±  |0.0118|
|     |       |strict-match    |     8|exact_match|↑  |0.7566|±  |0.0118|

### CODE REVIEW

File: vllm/attention/backends/hpu_attn.py [hunk 1]
```diff
@@ -542,6 +548,13 @@ def forward(
             block_list = attn_metadata.block_list if attn_metadata \
                 and attn_metadata.block_list is not None else None
 
+            common_args = self.common_attention_args(block_list, key_cache,
+                                                     value_cache,
+                                                     attn_metadata.block_size)
```
Comment (xuechendi) [inline vllm/attention/backends/hpu_attn.py:553]: Why move the common_args out?

File: vllm/model_executor/models/gemma3_mm.py [hunk 2]
```diff
@@ -610,6 +611,8 @@ def forward(self,
         # NOTE: In v1, inputs_embeds is always generated at model runner, this
         # condition is for v0 compatibility.
         elif inputs_embeds is None:
+            raise AssertionError("hpu_model_runner should be computing ",
+                                 "inputs_embeds")
```
Comment (xuechendi) [inline vllm/model_executor/models/gemma3_mm.py:615]: remove the line, let's minimize the modeling change, that will be difficult for rebasing and future support with vllm plugin since we can't touch modeling codes.

File: vllm/model_executor/models/gemma3_mm.py [hunk 3]
```diff
@@ -639,58 +642,53 @@ def prepare_attn_masks(
...
+        if self.sliding_window is not None:
+            # Create a local causal mask with sliding window (1024).
+            local_attn_mask = torch.ones_like(global_attn_mask)
+            local_attn_mask = torch.tril(local_attn_mask,
+                                         diagonal=-self.sliding_window)
+            local_attn_mask = torch.where(local_attn_mask == 0,
+                                          global_attn_mask, float("-inf"))
+
+        kwargs["global_attn_masks"] = global_attn_mask
+        kwargs["local_attn_masks"] = local_attn_mask
```
Comment (xuechendi) [inline vllm/model_executor/models/gemma3_mm.py:691]: same here, minimize the change to modeling codes. if the change is due to loop seq_lens is not correct supported by hpu  graph, let's add is_hpu() and with minimum change such as 

so we only introduce few lines to modeling
```
if is_hpu():
    seq_lens = list(input_ids.shape[1])

# use original codes:
for seq_len in seq_lens:
     end_idx = start_idx + seq_len
     ...
```

File: vllm/model_executor/models/utils.py [hunk 4]
```diff
@@ -394,6 +394,9 @@ def _merge_multimodal_embeddings(
     if current_platform.is_hpu():
         htcore.mark_step()
         flattened = _flatten_embeddings(multimodal_embeddings)
+        #TODO Check if it's dynamic - maybe torch.where?
+        #however multimodal_embeddings is a list of varying length
+        #torch.where might be faster than boolean indexing?
```
[the remark this answers is not in the document]
Comment (xuechendi) [reply vllm/model_executor/models/utils.py:399]: better to revert this file. We want to minimize the changes to general modules.

File: vllm/worker/hpu_model_runner.py [hunk 5]
```diff
@@ -354,6 +363,42 @@ def __init__(self, model, vllm_config, layer_names, is_causal, sampler):
             self.model.visual = htorch.hpu.wrap_in_hpu_graph(
                 self.model.visual, disable_tensor_cache=True)
 
+        # TODO : right now just enabling it keeping gemma3 in mind
+        if htorch.utils.internal.is_lazy() and is_gemma3(self.model):
+            logger.info("[Multimodal] Wrapping Visual Model")
+            self.model.vision_tower = htorch.hpu.wrap_in_hpu_graph(
+                self.model.vision_tower, disable_tensor_cache=True)
+            self.model.multi_modal_projector = htorch.hpu.wrap_in_hpu_graph(
+                self.model.multi_modal_projector, disable_tensor_cache=True)
```
Comment (xuechendi) [inline vllm/worker/hpu_model_runner.py:372]: is this part related to sliding window? What happened if we don't wrap hpu separately?

**Comment by xuechendi** [comment]
@madamczyk-intel , please take a look of hpu_model_runner.py changes

File: vllm/attention/backends/hpu_attn.py (hunk 1 above)
Comment (jiminha) [reply vllm/attention/backends/hpu_attn.py:553]: I will move it back. I was going to add additional parameter for left padding argument, but doesn't seem that we need it.

File: vllm/model_executor/models/gemma3.py [hunk 6]
```diff
@@ -233,44 +233,34 @@ def naive_attn_with_masks(
...
         num_queries_per_kv = self.num_heads // self.num_kv_heads
-        k = k.view(-1, self.num_kv_heads, self.head_dim)
-        k = k.repeat_interleave(num_queries_per_kv, dim=-2)
-        v = v.view(-1, self.num_kv_heads, self.head_dim)
-        v = v.repeat_interleave(num_queries_per_kv, dim=-2)
+        query = q.view(-1, s, self.num_heads, self.head_dim)
+        key = k.view(-1, s, self.num_kv_heads, self.head_dim)
+        key = key.repeat_interleave(num_queries_per_kv, dim=-2)
+        value = v.view(-1, s, self.num_kv_heads, self.head_dim)
+        value = value.repeat_interleave(num_queries_per_kv, dim=-2)
```
[the remark this answers is not in the document]
Comment (jiminha) [reply vllm/model_executor/models/gemma3.py:243]: @maktukmak  Could you please comment on this? I think this is part of the static mask change?

File: vllm/model_executor/models/gemma3.py [hunk 7]
```diff
@@ -233,44 +233,34 @@ def naive_attn_with_masks(
...
+        output = F.scaled_dot_product_attention(
+            query,
+            key,
+            value,
+            attn_masks,
+            self.scaling,
+        )
+
+        out = output.transpose(1, 2).flatten(-2, -1)
+
```
[the remark this answers is not in the document]
Comment (jiminha) [reply vllm/model_executor/models/gemma3.py:263]: @maktukmak  to comment further

File: vllm/model_executor/models/gemma3_mm.py (hunk 2 above)
Comment (jiminha) [reply vllm/model_executor/models/gemma3_mm.py:615]: will do

File: vllm/model_executor/models/gemma3_mm.py (hunk 3 above)
Comment (jiminha) [reply vllm/model_executor/models/gemma3_mm.py:691]: @maktukmak  to comment further

File: vllm/model_executor/models/gemma3.py [hunk 8]
```diff
@@ -233,44 +233,34 @@ def naive_attn_with_masks(
         out: torch.Tensor,
         **kwargs,
     ) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (jiminha) [reply vllm/model_executor/models/gemma3.py:242]: I combined code to just use hpu_attn with appropriate mask , so naive_attn_with_masks() will not be called in HPU .

File: vllm/model_executor/models/gemma3.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (libinta) [reply vllm/model_executor/models/gemma3.py:270]: @jiminha  please check my PR and backport,

File: vllm/worker/hpu_model_runner.py (hunk 5 above)
Comment (libinta) [reply vllm/worker/hpu_model_runner.py:372]: this is not part of sliding window, its for vision tower bucket handle

File: vllm/model_executor/models/gemma3.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jiminha) [reply vllm/model_executor/models/gemma3.py]: Removed.

**Comment by jiminha** [comment]
> @jiminha , please add UT for slidingwindow test

@xuechendi  UT is added for multimodal use case, and gsm8k longer prompts accuracy.

File: vllm/model_executor/models/gemma3_mm.py [hunk 9]
```diff
@@ -566,6 +573,7 @@ def _process_image_input(
             self.vision_tower,
             pixel_values,
         )
+
```
Comment (michalkuligowski) [inline vllm/model_executor/models/gemma3_mm.py:576]: please remove new line

File: vllm/model_executor/models/gemma3_mm.py [hunk 10]
```diff
@@ -639,30 +647,42 @@ def prepare_attn_masks(
...
+            start_idices = (positions == 0).cpu().nonzero()
+            num_seqs = len(start_idices)
+            for i in range(num_seqs):
+                start_idx = start_idices[i].item()
+                if i < num_seqs - 1:
+                    end_idx = start_idices[i + 1].item()
+                else:
+                    end_idx = len(input_ids)
+                seq_lens.append(end_idx - start_idx)
+                kwargs["seq_lens"] = seq_lens
```
Comment (michalkuligowski) [inline vllm/model_executor/models/gemma3_mm.py:670]: Wrong intendation.
kwargs["seq_lens"] is assigned outside the loop.

File: vllm/model_executor/models/utils.py [hunk 11]
```diff
@@ -457,11 +457,11 @@ def merge_multimodal_embeddings(
...
-    
-    ``placeholder_token_id`` can be a list of token ids (e.g, token ids 
-    of img_start, img_break, and img_end tokens) when needed: This means 
-    the order of these tokens in the ``input_ids`` MUST MATCH the order of 
-    their embeddings in ``multimodal_embeddings`` since we need to 
+
+    ``placeholder_token_id`` can be a list of token ids (e.g, token ids
+    of img_start, img_break, and img_end tokens) when needed: This means
+    the order of these tokens in the ``input_ids`` MUST MATCH the order of
+    their embeddings in ``multimodal_embeddings`` since we need to
```
Comment (michalkuligowski) [inline vllm/model_executor/models/utils.py:464]: Why are those changes in the documentation needed?

File: vllm/worker/hpu_model_runner.py [hunk 12]
```diff
@@ -327,7 +343,7 @@ def __init__(self, model, vllm_config, is_causal, sampler):
     def _get_rotary_embedding_module(self, model: torch.nn.Module):
         """
         Dynamically get the RotaryEmbedding layer in the model.
-        This function will recursively search through the module 
+        This function will recursively search through the module
```
Comment (michalkuligowski) [inline vllm/worker/hpu_model_runner.py:346]: Another empty line is deleted, do you have some whitespace rule in the git? Lets not change that

File: vllm/worker/hpu_model_runner.py [hunk 13]
```diff
@@ -416,37 +460,99 @@ def _set_attn_bias(self, attn_metadata, batch_size, seq_len, device,
...
+        tensor = torch.full((batch_size, 1, seq_len, seq_len),
+                            device=device,
+                            dtype=dtype,
+                            fill_value=1)
+        mask = torch.tril(tensor, diagonal=shift)
+        mask = torch.triu(mask, diagonal=shift - window_size + 1)
+        attn_bias = torch.log(mask)
+        '''
+        #TODO:causal + window size + q_len : accuracy and perf issue.
+        #Further need to be optimized when custom kernel is available.
```
Comment (michalkuligowski) [inline vllm/worker/hpu_model_runner.py:483]: Empty spaces needed after hash sign

File: vllm/worker/hpu_model_runner.py [hunk 14]
```diff
@@ -416,37 +460,99 @@ def _set_attn_bias(self, attn_metadata, batch_size, seq_len, device,
...
+            dtype=dtype, fill_value=1)
+        mask = torch.tril(tensor, diagonal=shift)
+        len_mask = torch.arange(0, seq_len, device=device,
+            dtype=torch.int32).view(seq_len,1)
+        len_mask = len_mask.ge(query_lens_t.unsqueeze(-1)).view(batch_size,
+            1, seq_len, 1)
+        len_mask = torch.where(len_mask == False, 1, 0)
+        mask = mask.logical_and(len_mask)
+        mask = torch.triu(mask, diagonal=shift - window_size + 1)
+        attn_bias =torch.where(mask,0, -math.inf)
```
Comment (michalkuligowski) [inline vllm/worker/hpu_model_runner.py:497]: please remove unused code

File: vllm/worker/hpu_model_runner.py [hunk 15]
```diff
@@ -416,37 +460,99 @@ def _set_attn_bias(self, attn_metadata, batch_size, seq_len, device,
...
             # all negative values to 0 (block_mapping) or bs (block_groups)
-            block_groups = metadata.block_groups.to(torch.long)
+            block_groups = block_groups.to(torch.long)
             block_mapping = torch.nn.functional.relu(block_groups)
             block_mapping = torch.nn.functional.one_hot(block_mapping,
                                                         num_classes=batch_size)
             oob_values = block_groups.lt(0)
             block_mapping.masked_fill_(oob_values.unsqueeze(-1), 0)
             block_groups.masked_fill_(oob_values, batch_size)
+            if not is_window_block:
```
Comment (michalkuligowski) [inline vllm/worker/hpu_model_runner.py:535]: please revers condition, its easier to read when its:
if is_window_block:

File: vllm/worker/hpu_model_runner.py [hunk 16]
```diff
@@ -416,37 +460,99 @@ def _set_attn_bias(self, attn_metadata, batch_size, seq_len, device,
...
+                                                "TrimmedAttentionMetadata",
+                                                block_groups=block_groups)
+            else:
+                metadata = custom_tuple_replace(
+                    metadata,
+                    "TrimmedAttentionMetadata",
+                    window_block_groups=block_groups)
+
+        block_mapping = block_mapping.to(dtype)
+        if not is_window_block:
```
Comment (michalkuligowski) [inline vllm/worker/hpu_model_runner.py:546]: as above, please reverse the condition

File: vllm/worker/hpu_model_runner.py [hunk 17]
```diff
@@ -416,37 +460,99 @@ def _set_attn_bias(self, attn_metadata, batch_size, seq_len, device,
...
+        attn_bias =torch.where(mask,0, -math.inf)
+        '''
+        attn_metadata = prefill_metadata._replace(window_attn_bias=attn_bias)
+
+        return attn_metadata
+
+    def _set_block_mapping(self, metadata, batch_size, device, dtype,
+                           is_window_block):
+
+        if not is_window_block:
```
Comment (michalkuligowski) [inline vllm/worker/hpu_model_runner.py:506]: please revers condition, its easier to read when its:
if is_window_block:

**Review by michalkuligowski — DISMISSED** [review]

File: vllm/worker/hpu_model_runner.py (hunk 13 above)
Comment (jiminha) [reply vllm/worker/hpu_model_runner.py]: Removed this whole block

File: vllm/worker/hpu_model_runner.py (hunk 14 above)
Comment (jiminha) [reply vllm/worker/hpu_model_runner.py]: Done

**Review by xuechendi — APPROVED** [review]

File: vllm/model_executor/models/gemma3_mm.py (hunk 10 above)
Comment (jiminha) [reply vllm/model_executor/models/gemma3_mm.py]: Done
